lib/MySQL_Thread.cpp

```diff
@@ -3558,90 +3271,35 @@ MySQL_Session * MySQL_Thread::create_new_session_and_client_data_stream(int _fd)
-		free(sess->client_myds->myconn->options.sql_auto_is_null);
-	}
-	sess->client_myds->myconn->options.sql_auto_is_null=strdup(mysql_thread___default_sql_auto_is_null);
-
-	uint32_t sql_select_limit_int=SpookyHash::Hash32(mysql_thread___default_sql_select_limit,strlen(mysql_thread___default_sql_select_limit),10);
-	sess->client_myds->myconn->options.sql_select_limit_int = sql_select_limit_int;
-	if (sess->client_myds->myconn->options.sql_select_limit) {
-		free(sess->client_myds->myconn->options.sql_select_limit);
-	}
-	sess->client_myds->myconn->options.sql_select_limit=strdup(mysql_thread___default_sql_select_limit);
-
-	uint32_t sql_safe_updates_int=SpookyHash::Hash32(mysql_thread___default_sql_safe_updates,strlen(mysql_thread___default_sql_safe_updates),10);
-	sess->client_myds->myconn->options.sql_safe_updates_int = sql_safe_updates_int;
-	if (sess->client_myds->myconn->options.sql_safe_updates) {
-		free(sess->client_myds->myconn->options.sql_safe_updates);
-	}
-	sess->client_myds->myconn->options.sql_safe_updates=strdup(mysql_thread___default_sql_safe_updates);
-
-	uint32_t collation_connection_int=SpookyHash::Hash32(mysql_thread___default_collation_connection,strlen(mysql_thread___default_collation_connection),10);
-	sess->client_myds->myconn->options.collation_connection_int = collation_connection_int;
-	if (sess->client_myds->myconn->options.collation_connection) {
-		free(sess->client_myds->myconn->options.collation_connection);
-	}
-	sess->client_myds->myconn->options.collation_connection=strdup(mysql_thread___default_collation_connection);
-
-	uint32_t net_write_timeout_int=SpookyHash::Hash32(mysql_thread___default_net_write_timeout,strlen(mysql_thread___default_net_write_timeout),10);
-	sess->client_myds->myconn->options.net_write_timeout_int = net_write_timeout_int;
-	if (sess->client_myds->myconn->options.net_write_timeout) {
-		free(sess->client_myds->myconn->options.net_write_timeout);
-	}
-	sess->client_myds->myconn->options.net_write_timeout=strdup(mysql_thread___default_net_write_timeout);
-
-	uint32_t max_join_size_int=SpookyHash::Hash32(mysql_thread___default_max_join_size,strlen(mysql_thread___default_max_join_size),10);
-	sess->client_myds->myconn->options.max_join_size_int = max_join_size_int;
-	if (sess->client_myds->myconn->options.max_join_size) {
-		free(sess->client_myds->myconn->options.max_join_size);
+	for (int i=0; i<SQL_NAME_LAST; i++) {
+		if (i == SQL_CHARACTER_SET || i == SQL_CHARACTER_SET_RESULTS ||
+			i == SQL_CHARACTER_SET_CONNECTION || i == SQL_CHARACTER_SET_CLIENT ||
+			i == SQL_CHARACTER_SET_DATABASE) {
+			const MARIADB_CHARSET_INFO *ci = NULL;
+			ci = proxysql_find_charset_name(mysql_thread___default_variables[i]);
+			if (!ci) {
+				proxy_error("Cannot find character set for name [%s]. Configuration error. Check [%s] global variable. Using character set 33.\n", 
+						mysql_thread___default_variables[i], mysql_tracked_variables[i].internal_variable_name);
+				assert(0);
```

Comment:
A fresh installation crashes here while a client tried to connect (before the handshake).

Reply:
This needs enhancement in `ProxySQL_Admin::flush_mysql_variables___database_to_runtime()` .
Before loading all the variables, it can scan them all and fix incompatibilities.
For example:
* if a charset is set to something not existing, set to the hardcoded default;
* if a collation is configured to a not existing value, change it to the default for the configured charset
* if the configured collation is of a different charset, change it to the default collation for the configured charset

Reply:
Fixed, please have a look.